[core] Fix race condition b/w object eviction & repinning for recovery.

israbbani opened the pull request:
This PR has two components:

### Fixing the Race Condition b/w Object Eviction and Repining for Object Recovery  

The sequence of events documents the life of an object called A. 

1. Owner sends object eviction notifications to all locations as part of Object A going out of scope. (done through worker<->worker pubsub) 
2. Workers that have a copy of Object A mark it for asynchronous deletion.
3. Owner decides to recover the object and repin a copy of object A that has already been marked for asynchronous deletion.
4. Owner submits a task that has Object A as an argument.
5. The task executor tries to fetch Object A and the pull manager hangs forever.

The fix is to check to see if an object is pending deletion before pinning it. If it is, the owner cannot pin the object and has to either try a different location or trigger reconstruction by resubmitting the task that created the object.

### (The Joys of) Writing a Unit Test in Node Manager

There are a few problems with writing unit tests in the NodeManager

1. Despite heroics from @rueian and @dayshah recently, the NodeManager still had a few dependencies that needed to be put behind interfaces so fakes could be injected.
2. The gRPC handlers are all private to the class so they cannot be called directly. Instead the unit tests need to create a gRPC client. 
3. The NodeManager has a lot of dependencies and the dependencies have interdependencies so we currently dedicate 200 lines to constructing a test fixture.
4. There is a lack of usable fakes e.g. for the PlasmaClient for writing unit tests. 

### Follow Ups
1. Create a shim between the NodeManager and it's gRPC handlers so they can be unit tested directly without creating gRPC clients.
2. Create a utility function that lets you construct a NodeManager for testing optionally overriding interesting dependencies (otherwise they will default to a Noop fake).
3. Move the fakes created in this PR into their own (more complete implementations).
4. Create a Fake for SubscriberInterface.

Commit:
[core] Fix race condition b/w object eviction & repinning for recovery.

The sequence of events:

1. The owner sends object eviction notifications to all locations as part
of an object going out of scope.
2. The remote copies are deleted asynchronously.
3. The owner can decide to recover the object and repin the copy
that has already been marked for asynchronous deletion.
4. The pull manager hangs forever.

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
formatting

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Merge branch 'master' into irabbani/CORE-1618

Commit:
Merge branch 'master' into irabbani/CORE-1618

Commit:
Merge branch 'master' into irabbani/CORE-1618

Commit:
[core] Making node manager more testable by
1. Adding an interface for LocalObjectManager
2. Adding a unit test

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
Merge branch 'master' into irabbani/CORE-1618

Signed-off-by: Ibrahim Rabbani <irabbani@anyscale.com>

Commit:
format

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
merge conflicts

Signed-off-by: irabbani <irabbani@anyscale.com>

israbbani commented on src/ray/raylet/node_manager.cc:
```diff
@@ -2490,11 +2491,16 @@ void NodeManager::HandlePinObjectIDs(rpc::PinObjectIDsRequest request,
+      // Therefore, a new object cannot be marked for deletion while this function is
+      // executing.
+      if (*result_it == nullptr ||
+          local_object_manager_.ObjectPendingDeletion(*object_id_it)) {
```
Major logic change.

dayshah commented on src/ray/raylet/local_object_manager_interface.h:
```diff
@@ -0,0 +1,74 @@
+
+};  // namespace raylet
+
+};  // namespace ray
```
i'd prefer just putting this in local_object_manager.h

israbbani replied:
I put it in a separate file so we can have a separate build target for the interface. Most of our other build targets can link to this and it'll speed up our compile times.

dayshah commented on src/ray/raylet/test/node_manager_test.cc:
```diff
@@ -39,6 +43,149 @@ using ::testing::Return;
+  absl::flat_hash_set<ObjectID> objects_ids_in_plasma_;
+  absl::flat_hash_map<ObjectID, std::pair<std::vector<uint8_t>, std::vector<uint8_t>>>
+      objects_in_plasma_;
+};
```
imo prefer mocks in mock/ to no-op fakes, more similar to the rest of the code too

israbbani replied:
I disagree. I am hoping to remove as much of our mock code as possible and replace it with fake implementations. Mocks encourage intercepting calls and break encapsulation and overfit the tests to the implementations rather than the API.

dayshah replied:
I conceptually agree, intercepting calls and actually using the mocks everywhere makes the tests brittle. The reason i prefer the mocks is because they give you the default no-op implementations by default and give you the option of overriding / intercepting whatever's necessary. It'll generally result in less verbose test files and it's up to the reviewers to make sure people aren't writing bad tests that overfit to implementations.

I'm open to going with shared no-op implementations instead of mocks. But that's effectively the same thing except that devs will be more discouraged to mock bc it means writing more code?

But this is probably a broader team decision + a shift in the way we do things now. cc @jjyao + @edoakes

dayshah commented on src/ray/raylet/test/node_manager_test.cc:
```diff
@@ -644,6 +770,61 @@ TEST_F(NodeManagerTest, TestDetachedWorkerIsKilledByFailedNode) {
+
+    grpc::ClientContext context;
+    auto stub = rpc::NodeManagerService::NewStub(channel);
+    auto status = stub->PinObjectIDs(&context, request, &reply);
```
just make the test a friend and call HandlePinObjectIDs with the request and reply, this would simplify things a lot and this complexity in a test isn't worth it

Also idk if handlers should really be private, this is a node manager only thing. Gcs handlers are public and that makes sense... the handlers are expected to be triggered extrenally.

israbbani replied:
Making the test a friend breaks encapsulation. Tests should be written against the public API of the class as much as possible. The short-term pain of making code testable is worth the long-term gain of not having brittle tests and it'll force us to improve code quality. 

Agreed that the handlers should be public. I'm planning on changing this in a follow up.

dayshah replied:
can we just make them public here and avoid merging the grpc logic in test code

dayshah commented on src/ray/raylet/test/node_manager_test.cc:
```diff
@@ -644,6 +770,61 @@ TEST_F(NodeManagerTest, TestDetachedWorkerIsKilledByFailedNode) {
+
+  std::thread io_thread{[&] {
+    auto work_guard = boost::asio::make_work_guard(io_service_);
+    io_service_.run();
```
ideally shouldn't just let the io_service_ run on its own thread in a test. You can just call run_one on the test service as needed.

israbbani replied:
> ideally shouldn't just let the io_service_ run on its own thread in a test. You can just call run_one on the test service as needed.

Super agree. Can you say a little more about "You can just call run_one on the test service as needed."?

edoakes replied:
Super super agree. We should ban this.

Here is a test I wrote awhile back that does this: https://github.com/ray-project/ray/blob/5957aab7adff05c010a72b784e7a407981707e8e/src/ray/common/test/client_connection_test.cc#L126

israbbani replied:
Turns out I was being lazy. I don't think I need the io_service to be running since I'm calling a function that doesn't require it (and none of the periodical_runner_ stuff depends on it).

dayshah reviewed (commented):
> Create a utility function that lets you construct a NodeManager for testing optionally overriding interesting dependencies (otherwise they will default to a Noop fake).

Instead of this we should just have mocks and you can override implementations of a mock per test or even per invocation in a test. More flexible and less code

Also node_manager_test failing build bc warnings

israbbani commented:
> Also node_manager_test failing build bc warnings

This is interesting. CI builds with clang. I'm using gcc on my devbox. I'll change it to clang. Good catch!

Commit:
Fixing clang build

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
A few updates to the unit test:
- making HandlePinObjectIDs public so it can be tested directly
- removing io_service_

Signed-off-by: irabbani <irabbani@anyscale.com>

Commit:
removing io_service.stop()

Signed-off-by: irabbani <irabbani@anyscale.com>

dayshah reviewed (approved):
🚢
